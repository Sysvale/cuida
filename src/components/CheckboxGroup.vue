<template>
	<div class="checkbox__container">
		<label class="checkbox__label">
			{{ label }}
			<span
				v-if="required"
				class="checkbox__required"
			>
				*
			</span>
		</label>
		<div
			v-for="option in options"
			:key="option.value"
			:class="resolveCheckboxClass(option.value)"
			@click="handleCheckboxClick(option.value)"
		>
			<cds-checkbox
				:id="option.value"
				:model-value="internalValue.includes(option.value)"
				:label="option.label"
				:disabled="disabled"
				:variant="variant"
			/>
		</div>
		<div
			v-if="isInvalid && !disabled"
			class="checkbox__error-message"
		>
			{{ errorMessage }}
		</div>
	</div>
</template>

<script setup>
import { ref, onMounted, computed, watch } from 'vue';
import CdsCheckbox from './Checkbox.vue';
import variantClassResolver from '../utils/methods/variantClassResolver';
import variantValidator from '../utils/validators/variant';

const props = defineProps({
	/**
	* Prop usada como v-model para monitorar as seleções do CheckboxGroup
	*/
	modelValue: {
		type: Array,
		default: () => ([]),
	},
	/**
	* Opções disponíveis para seleção na lista a ser renderizada pelo componente. Deve ser formada por objetos contendo `label` e `value`.
	*/
	options: {
		type: Array,
		default: () => ([]),
	},
	/**
	* Etiqueta do componente.
	*/
	label: {
		type: String,
		default: 'Label',
	},
	/**
	 * Exibe asterisco de obrigatório (obs.: não faz a validação)
	 */
	required: {
		type: Boolean,
		default: false,
	},
	/**
	* A variante do componente. São 10 variantes: 'teal', 'green', 'blue',
	* 'indigo', 'violet', 'pink', 'red', 'orange', 'amber' e 'dark'.
	*/
	variant: {
		type: String,
		default: 'green',
		validator: variantValidator,
	},
	/**
	* Controla o status checkbox
	*/
	disabled: {
		type: Boolean,
		default: false,
	},
	/**
	 * Especifica o estado do CheckboxGroup. As opções são 'default', 'valid' e 'invalid'.
	 */
	state: {
		type: String,
		default: 'default',
	},
	/**
	 * Especifica a mensagem de erro, que será exibida caso o estado seja inválido
	 */
	errorMessage: {
		type: String,
		default: 'Valor inválido',
	},
});

const emits = defineEmits(['update:modelValue']);

const internalValue = ref([]);
const checkboxControl = ref(0);

const isInvalid = computed(() => props.state === 'invalid');

watch(internalValue, () => checkboxControl.value += 1);

onMounted(() => {
	internalValue.value = props.modelValue;
});

function resolveCheckboxClass(selectedOption) {
	if (props.disabled) {
		return 'checkbox__item--disabled';
	}

	const variantClass = isInvalid.value
		? 'checkbox__item--error'
		: variantClassResolver('checkbox__item', props.variant);

	return internalValue.value.includes(selectedOption)
		? `checkbox__item--selected ${variantClass}`
		: `checkbox__item ${isInvalid.value ? variantClass : ''}`;
}

function handleCheckboxClick(selectedOption) {
	if (props.disabled) {
		return;
	}

	if (internalValue.value.includes(selectedOption)) {
		internalValue.value = internalValue.value.filter((option) => option !== selectedOption);
		emits('update:modelValue', internalValue.value);
		return;
	}

	internalValue.value.push(selectedOption);
	emits('update:modelValue', internalValue.value);
}
</script>

<style lang="scss" scoped>
@use '../assets/sass/tokens/index' as tokens;

.checkbox {
	&__container {
		display: flex;
		flex-direction: column;
		gap: spacer(2);
	}

	&__label {
		@include tokens.button-2;
		margin: mb(2);
	}

	&__item {
		cursor: pointer;
		padding: pYX(4, 3);
		@include tokens.body-2;
		border: 1px solid tokens.$n-50;
		border-radius: tokens.$border-radius-extra-small;

		@include tokens.variantResolver using ($color-name, $shade-50, $shade-100, $shade-200, $shade-300, $base-color, $shade-500, $shade-600) {
			border-color: $base-color !important;
		}

		&--error {
			border-color: tokens.$rc-600;
		}

		&--selected {
			@extend .checkbox__item;
		}

		&--disabled {
			@extend .checkbox__item;
			background-color: tokens.$n-10;
		}
	}

	&__required {
		color: tokens.$rc-600;
	}

	&__error-message {
		@include tokens.caption;
		color: tokens.$rc-600;
		margin: mt(1);
	}
}
</style>
