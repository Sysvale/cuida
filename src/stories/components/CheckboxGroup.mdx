import { Meta, Story, Props, Controls, Canvas, Preview } from '@storybook/blocks';
import CheckboxGroup from '../../components/CheckboxGroup.vue';
import * as CheckboxGroupStories from './CheckboxGroup.stories';

<Meta of={CheckboxGroupStories} />

export const Template = (args) => ({
	components: { CdsCheckboxGroup: CheckboxGroup },
	setup() {
		return { args };
	},
	data() {
		return {
			value: ['test-1'],
		};
	},
	template: /*html*/ `
		<cds-checkbox-group
			v-bind="args"
			@update:model-value="handleValueChange"
		/>
	`,
	methods: {
		handleValueChange(value) {
			console.info('v-model: ', value);
		},
	},
});

# CheckboxGroup

### CheckboxGroups são componentes de formulário que permitem seleções binárias e múltiplas.

***

<br />

## Quando usar:

* Quando precisar de um componente de formulário para seleções binárias com várias opções.

<br />

## Quando não usar:

* Quando for preciso que o efeito da interação com o Checkbox seja imediato. Checkboxes devem ser usados sempre em conjunto com um botão de submissão. Para contornar casos como esse, use Switches.

## Preview

<Canvas>
  <Story of={CheckboxGroupStories.CheckboxGroup} />
</Canvas>

<Controls />
