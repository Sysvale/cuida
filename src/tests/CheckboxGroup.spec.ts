import { describe, test, expect } from 'vitest';
import CheckboxGroup from '../components/CheckboxGroup.vue';
import CheckboxItem from '../components/Checkbox.vue';
import { mount } from '@vue/test-utils';
  
describe('CheckboxGroup', () => {
	test('renders correctly', async () => {
		const wrapper = mount(CheckboxGroup, {
			props: {
				modelValue: ['test-1'],
				label: 'Checkbox Group Test',
				options: [
					{
						value: 'test-1',
						label: 'Teste 1',
					},
					{
						value: 'test-2',
						label: 'Teste 2',
					},
				],
			},
		});
  
		expect(wrapper.html()).toMatchSnapshot();
	});

	test('checkbox options are rendered correctly', async () => {
		expect.assertions(2);

		const wrapper = await mount(CheckboxGroup, {
			props: {
				modelValue: ['test-1'],
				label: 'Checkbox Group Test',
				options: [
					{
						value: 'test-1',
						label: 'Teste 1',
					},
					{
						value: 'test-2',
						label: 'Teste 2',
					},
				],
			},
		});

		const checkboxArray = wrapper.findAllComponents(CheckboxItem);
		const checkedCheckbox = checkboxArray.find((checkbox) => checkbox.vm.label === 'Teste 1');

		expect(checkboxArray.length).toBe(2);
		expect(checkedCheckbox?.vm.modelValue).toBeTruthy();
	});
});
