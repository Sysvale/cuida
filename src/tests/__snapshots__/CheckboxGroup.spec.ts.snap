// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CheckboxGroup > renders correctly 1`] = `
"<div data-v-5505c754="" class="checkbox__container"><label data-v-5505c754="" class="checkbox__label">Checkbox Group Test
    <!--v-if-->
  </label>
  <div data-v-5505c754="" class="checkbox__item">
    <div data-v-75623c80="" data-v-5505c754="" class="cds-checkbox__container" id="test-1">
      <div data-v-75623c80="" class="cds-checkbox__input"><input data-v-75623c80="" id="test-1" type="checkbox" name="cds-checkbox-option" value="false"><label data-v-75623c80="" for="test-1" class="cds-checkbox__input--green cds-checkbox__input--unchecked"></label></div><label data-v-75623c80="" class="cds-checkbox__label" for="test-1">Teste 1</label>
    </div>
  </div>
  <div data-v-5505c754="" class="checkbox__item">
    <div data-v-75623c80="" data-v-5505c754="" class="cds-checkbox__container" id="test-2">
      <div data-v-75623c80="" class="cds-checkbox__input"><input data-v-75623c80="" id="test-2" type="checkbox" name="cds-checkbox-option" value="false"><label data-v-75623c80="" for="test-2" class="cds-checkbox__input--green cds-checkbox__input--unchecked"></label></div><label data-v-75623c80="" class="cds-checkbox__label" for="test-2">Teste 2</label>
    </div>
  </div>
  <!--v-if-->
</div>"
`;
